{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Import packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import io\n",
    "\n",
    "import scipy.stats as stats\n",
    "import pingouin as pg\n",
    "\n",
    "import random\n",
    "from random import randrange\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import argparse\n",
    "\n",
    "import Orthoscripts\n",
    "\n",
    "pd.options.mode.chained_assignment = None "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import genelists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Asterias rubens\n",
    "Astrub = Orthoscripts.readBED(\"Data/Genelists/Asterias.rubens.genelist.bed\")\n",
    "\n",
    "# Holothuria leucospilota\n",
    "Holleu = Orthoscripts.readBED(\"Data/Genelists/Holothuria.leucospilota.genelist.bed\")\n",
    "\n",
    "# Paracentrotus livides\n",
    "Parliv = Orthoscripts.readBED(\"Data/Genelists/Paracentrotus.lividus.genelist.bed\")\n",
    "\n",
    "# Branchiostoma lanceolatum\n",
    "Bralan = Orthoscripts.readBED(\"Data/Genelists/Branchiostoma.lanceolatum.genelist.bed\")\n",
    "\n",
    "# Branchiostoma floridae\n",
    "Braflo = Orthoscripts.readBED(\"Data/Genelists/Branchiostoma.floridae.genelist.bed\", 's')\n",
    "\n",
    "# Marthasterias glacialis\n",
    "Margla = Orthoscripts.readBED(\"Data/Genelists/Marthasterias.glacialis.genelist.bed\")\n",
    "\n",
    "# Pecten maximus\n",
    "Pecmax = Orthoscripts.readBED(\"Data/Genelists/Pecmax.genelist.bed\", 's')\n",
    "\n",
    "# Stichopus chloronotus\n",
    "Stichl = Orthoscripts.readBED(\"Data/Genelists/Stichopus.chloronotus.genelist.bed\")\n",
    "\n",
    "# Amphiura filiformis \n",
    "Ampfil = Orthoscripts.readBED(\"Data/Genelists/Amphiura.filiformis.genelist.bed\")\n",
    "\n",
    "# Ephydatia muelleri\n",
    "Ephmue = Orthoscripts.readBED(\"Data/Genelists/Ephmue.genelist.bed\", 's')\n",
    "\n",
    "# Ancestor \n",
    "AniAnc = Orthoscripts.readBED(\"Data/Genelists/AniAnc.genelist.bed\", 's')\n",
    "BilAnc = Orthoscripts.readBED(\"Data/Genelists/BilAnc.genelist.bed\", 's')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import ortholog files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import orthologs\n",
    "Astrub_Holleu = np.loadtxt(\"Data/Orthologs/Asterias.rubens+Holothuria.leucospilota.txt\", dtype = \"str\")\n",
    "\n",
    "Astrub_Parliv = np.loadtxt(\"Data/Orthologs/Asterias.rubens+Paracentrotus.lividus.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Parliv = np.loadtxt(\"Data/Orthologs/Holothuria.leucospilota+Paracentrotus.lividus.txt\", dtype = \"str\")\n",
    "\n",
    "Margla_Bralan = np.loadtxt(\"Data/Orthologs/Marthasterias.glacialis+Branchiostoma.lanceolatum.txt\", dtype = \"str\")\n",
    "\n",
    "Margla_Pecmax = np.loadtxt(\"Data/Orthologs/Marthasterias.glacialis+Pecten.maximus.txt\", dtype = \"str\")\n",
    "\n",
    "Margla_Stichl = np.loadtxt(\"Data/Orthologs/Marthasterias.glacialis+Stichopus.chloronotus.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Bralan = np.loadtxt(\"Data/Orthologs/Pecten.maximus+Branchiostoma.lanceolatum.txt\", dtype = \"str\")\n",
    "\n",
    "Stichl_Bralan = np.loadtxt(\"Data/Orthologs/Stichopus.chloronotus+Branchiostoma.lanceolatum.txt\", dtype = \"str\")\n",
    "\n",
    "Stichl_Pecmax = np.loadtxt(\"Data/Orthologs/Stichopus.chloronotus+Pecten.maximus.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Holleu = np.loadtxt(\"Orthology pipeline/orthologs/Pecmax+Holleu_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Bralan = np.loadtxt(\"Orthology pipeline/orthologs/Holleu+Bralan_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Bralan = np.loadtxt(\"Orthology pipeline/orthologs/Pecmax+Bralan_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Braflo = np.loadtxt(\"Orthology pipeline/orthologs/Pecmax+Braflo_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Braflo = np.loadtxt(\"Orthology pipeline/orthologs/Holleu+Braflo_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Ampfil = np.loadtxt(\"Data/Orthologs/Holothuria.leucospilota+Amphiura.filiformis.txt\", dtype = \"str\")\n",
    "\n",
    "Braflo_Ephmue = np.loadtxt(\"Orthology pipeline/orthologs/Braflo+Ephmue_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Ephmue = np.loadtxt(\"Orthology pipeline/orthologs/Holleu+Ephmue_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Ephmue = np.loadtxt(\"Orthology pipeline/orthologs/Pecmax+Ephmue_sensitive.txt\", dtype = \"str\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sorting out the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "Astrub = Astrub.loc[Astrub['Chromosome'].str.contains('chr')]\n",
    "Bralan = Bralan.loc[Bralan['Chromosome'].str.contains('BFL_')]\n",
    "Braflo = Braflo.loc[Braflo['Chromosome'].str.contains('BFL_')]\n",
    "Pecmax = Pecmax.loc[Pecmax['Chromosome'].str.contains('PYE_')]\n",
    "Ephmue = Ephmue.loc[Ephmue['Chromosome'].str.contains('EMU_')]\n",
    "\n",
    "# Ephmue genelist: remove suffix\n",
    "Ephmue['Name'] = Ephmue['Name'].str.rsplit('.t1').str.get(0)\n",
    "\n",
    "# Parliv genelist: select chromosomal scaffolds\n",
    "Parliv = Orthoscripts.unscaff(Parliv, 100)\n",
    "Ampfil = Orthoscripts.unscaff(Ampfil, 100)\n",
    "Ephmue = Orthoscripts.unscaff(Ephmue, 600)\n",
    "\n",
    "Astrub_Parliv = Orthoscripts.orthFix(Astrub_Parliv, 'B', 'Parliv_', 1)\n",
    "Margla_Bralan = Orthoscripts.orthFix(Margla_Bralan, 'A', '.1', 0)\n",
    "Margla_Stichl = Orthoscripts.orthFix(Margla_Stichl, 'A', '.1', 0)\n",
    "Margla_Stichl = Orthoscripts.orthFix(Margla_Stichl, 'B', '.1', 0)\n",
    "Margla_Pecmax = Orthoscripts.orthFix(Margla_Pecmax, 'B', '.1', 0)\n",
    "Holleu_Ampfil = Orthoscripts.orthFix(Holleu_Ampfil, 'B', '.1', 0)\n",
    "Holleu_Bralan = Orthoscripts.orthFix(Holleu_Bralan, 'B', '_', 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### \n",
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = Orthoscripts.orthofind(Braflo, Ephmue, Braflo_Ephmue)\n",
    "Orthoscripts.orthoplot(data, 'Amphioxus', 'Sponge', 'A', 'B')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------\n",
    "\n",
    "### Test simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulator(Nchr = 20, Ngene = 100, Nevents = 10):\n",
    "\n",
    "    # Make ancestor genome\n",
    "    def makeancestor(Nchr, Ngene):\n",
    "        ancestor = pd.DataFrame(columns = ['Chromosome'])\n",
    "        for i in range(Nchr):\n",
    "            row = {'Chromosome' : (i + 1)}\n",
    "            for i in range(Ngene):\n",
    "                    ancestor = pd.concat([ancestor, pd.DataFrame([row])], ignore_index = True)\n",
    "        ancestor['Name'] = (ancestor.reset_index().index + 1)\n",
    "        # ancestor['Name'] = 'g_' + ancestor['Name'].astype(str)\n",
    "\n",
    "        return ancestor\n",
    "\n",
    "    # Dummy BED files :: type 'anc' for ancestor, 'des' for descendant\n",
    "    def dummyBED(genome, type):\n",
    "        if type == 'anc':\n",
    "            genome['Chromosome'] = 'AncChr' + genome['Chromosome'].astype(str)\n",
    "            genome['Name'] = 'ancg_' + genome['Name'].astype(str)\n",
    "            \n",
    "        if type == 'des':\n",
    "            genome['Chromosome'] = 'Chr' + genome['Chromosome'].astype(str)\n",
    "            genome['Name'] = 'g_' + genome['Name'].astype(str)\n",
    "        \n",
    "        genome['Start'] = np.arange(len(genome))\n",
    "        genome['End'] = np.arange(len(genome)) + 5\n",
    "        \n",
    "        genome = genome[['Chromosome', 'Start', 'End', 'Name']]\n",
    "            \n",
    "        return genome\n",
    "\n",
    "    # Dummy ortholog file\n",
    "    def dummyOrthologs(genome):\n",
    "        orthologs = pd.DataFrame()\n",
    "        \n",
    "        orthologs['Orthologs'] = np.arange(len(genome)) + 1\n",
    "        orthologs['speciesA'] = np.arange(len(genome)) + 1\n",
    "        orthologs['speciesB'] = np.arange(len(genome)) + 1\n",
    "        \n",
    "        orthologs['Orthologs'] = 'orthologs_' + orthologs['Orthologs'].astype(str)\n",
    "        orthologs['speciesA'] = 'ancg_' + orthologs['speciesA'].astype(str)\n",
    "        orthologs['speciesB'] = 'g_' + orthologs['speciesB'].astype(str)\n",
    "        \n",
    "        orthologs = orthologs.to_numpy()\n",
    "        \n",
    "        return orthologs\n",
    "\n",
    "    def mixing(genome, mixing):\n",
    "        genes = genome['Name'].to_numpy()\n",
    "        n = len(genes)\n",
    "        for i in range(int(mixing * n)):\n",
    "            g1, g2 = randrange(n), randrange(n)\n",
    "            genes[g2], genes[g1] = genes[g1], genes[g2]\n",
    "\n",
    "            genome['Name'] = genes\n",
    "            # genome['Chromosome'] = f'{fuse1}x{fuse2}'\n",
    "            \n",
    "    def fusion(genome, chr, mixing = 0):\n",
    "        '''\n",
    "        inputs: \n",
    "        ancestor : \n",
    "            df with chromosome name | gene name\n",
    "        mixing : \n",
    "            float between 0 and 1, where 1 implies \n",
    "            extreme mixing and 0 implies no mixing\n",
    "        '''\n",
    "        \n",
    "        # Randomly select two chromosomes to fuse\n",
    "        A = random.choice(chr)\n",
    "        B = random.choice(chr)\n",
    "        \n",
    "        chr = [x for x in chr if x not in (A, B)]\n",
    "        \n",
    "        if A == B: # Just so the same chromosome isn't selected twice\n",
    "            B = random.choice(chr)\n",
    "            if A == B:\n",
    "                B = random.choice(chr)\n",
    "\n",
    "        fusion = ancestor.loc[ancestor['Chromosome'].isin([A, B])]\n",
    "        \n",
    "        # Apply mixing if required\n",
    "        if mixing > 0:\n",
    "            genes = fusion['Name'].to_numpy()\n",
    "            n = len(genes)\n",
    "            for i in range(int(mixing * n)):\n",
    "                g1, g2 = randrange(n), randrange(n)\n",
    "                genes[g2], genes[g1] = genes[g1], genes[g2]\n",
    "\n",
    "            fusion['Name'] = genes\n",
    "            fusion['Chromosome'] = f'{A}x{B}'\n",
    "            log = f'FUS, AncChr{A} AncChr{B}, Chr{A}x{B}'\n",
    "            \n",
    "        else:\n",
    "            fusion['Chromosome'] = f'{A}+{B}'\n",
    "            log = f'FUS, AncChr{A} AncChr{B}, Chr{A}+{B}'\n",
    "        \n",
    "        # Remove the unfused chromosomes\n",
    "        genome.drop(genome[genome['Chromosome'].isin([A, B])].index, inplace = True)\n",
    "        genome = pd.concat([genome, fusion])\n",
    "        \n",
    "        return genome, log, chr\n",
    "\n",
    "    def fission(genome, chr):\n",
    "        # Randomly select a chromosome for fission\n",
    "        A = random.choice(chr)\n",
    "        fission = genome.loc[genome['Chromosome'] == A]\n",
    "        chr.remove(A)\n",
    "        \n",
    "        pos = random.choice(range(1, Ngene))\n",
    "\n",
    "        # Add the new chromosomes back into the genome\n",
    "        chr1 = fission.iloc[: pos]\n",
    "        chr1['Chromosome'] = f'{A}_1'\n",
    "        \n",
    "        chr2 = fission.iloc[pos :]\n",
    "        chr2['Chromosome'] = f'{A}_2'\n",
    "        \n",
    "        # Remove the fission chromosome from the genome\n",
    "        genome = pd.concat([genome, chr1, chr2])\n",
    "        genome = genome[genome.Chromosome != A]\n",
    "        \n",
    "        log = f'FIS, AncChr{A}, Chr{A}_1 Chr{A}_2'\n",
    "        \n",
    "        return genome, log, chr\n",
    "\n",
    "    def translocation(genome, chr):\n",
    "        # Randomly select two chromosomes for translocation\n",
    "        A = random.choice(chr)\n",
    "        B = random.choice(chr)\n",
    "        \n",
    "        if A == B: # Just so the same chromosome isn't selected twice\n",
    "            B = random.choice(chr)\n",
    "        \n",
    "        chr = [x for x in chr if x not in (A, B)]\n",
    "        \n",
    "        chrA = genome.loc[genome['Chromosome'] == A]\n",
    "        chrB = genome.loc[genome['Chromosome'] == B]\n",
    "        \n",
    "        # Randomly select two break point positions\n",
    "        posA = random.choice(range(1, Ngene))\n",
    "        posB = random.choice(range(1, Ngene))\n",
    "        \n",
    "        r = np.random.uniform()\n",
    "        if r <= 0.30:\n",
    "            # Join the fragments to form recombinant chromosomes\n",
    "            chr1 = pd.concat([chrA.iloc[: posA], chrB.iloc[posB :]])\n",
    "            chr1['Chromosome'] = f'{A};{B}'\n",
    "            chr2 = pd.concat([chrB.iloc[: posB], chrA.iloc[posA :]])\n",
    "            chr2['Chromosome'] = f'{B};{A}'\n",
    "            \n",
    "            log = f'TR, AncChr{A} AncChr{B}, Chr{A};{B} Chr{B};{A}'\n",
    "        \n",
    "        else:\n",
    "            chr1 = pd.concat([chrA.iloc[: posA]])\n",
    "            chr1['Chromosome'] = f'{A};'\n",
    "            chr2 = pd.concat([chrB, chrA.iloc[posA :]])\n",
    "            chr2['Chromosome'] = f'{B};{A}'\n",
    "        \n",
    "            log = f'TR, AncChr{A} AncChr{B}, Chr{B};{A}'\n",
    "            \n",
    "        # Remove the original chromosomes from the genome\n",
    "        genome = pd.concat([genome, chr1, chr2]).drop(genome[(genome['Chromosome'] == A) & (genome['Chromosome'] == B)].index)\n",
    "        return genome, log, chr\n",
    "\n",
    "    # Apply macro-rearrangements to the ancestor\n",
    "    ancestor = makeancestor(Nchr, Ngene)\n",
    "    chr = ancestor.Chromosome.unique().tolist()\n",
    "    speciesA = ancestor.copy()\n",
    "\n",
    "    events = []\n",
    "    for event in range(Nevents):\n",
    "        r = np.random.uniform()\n",
    "        \n",
    "        if r <= 0.30:\n",
    "            if len(ancestor) < 2: continue\n",
    "            speciesA, log, chr = fission(speciesA, chr)\n",
    "            events.append(log)\n",
    "        \n",
    "        elif r <= 0.45:\n",
    "            speciesA, log, chr = translocation(speciesA, chr)\n",
    "            events.append(log)\n",
    "        \n",
    "        elif r <= 0.70:\n",
    "            speciesA, log, chr = fusion(speciesA, chr)\n",
    "            events.append(log)\n",
    "        \n",
    "        elif r <= 0.99:\n",
    "            speciesA, log, chr = fusion(speciesA, chr, mixing = 0.5)\n",
    "            events.append(log)\n",
    "            \n",
    "        else:\n",
    "            # speciesA, log, chr = syntenyloss(speciesA, chr)\n",
    "            # events.append(log)\n",
    "            # print(log)\n",
    "            continue\n",
    "        \n",
    "    ancestor = dummyBED(ancestor, 'anc')\n",
    "    speciesA = dummyBED(speciesA, 'des')\n",
    "    orthologs = dummyOrthologs(ancestor)\n",
    "    \n",
    "    return ancestor, speciesA, orthologs, events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'FUS, AncChr7 AncChr3, Chr7x3 FUS, AncChr15 AncChr12, Chr15x12 TR, AncChr1 AncChr4, Chr1;4 Chr4;1 FIS, AncChr5, Chr5_1 Chr5_2 FIS, AncChr17, Chr17_1 Chr17_2 FIS, AncChr10, Chr10_1 Chr10_2 FUS, AncChr6 AncChr11, Chr6+11 FIS, AncChr8, Chr8_1 Chr8_2 FUS, AncChr14 AncChr16, Chr14+16 FUS, AncChr20 AncChr19, Chr20+19'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simancestor, simspeciesA, simorthologs, simevents = simulator()\n",
    "data = Orthoscripts.orthologies(simancestor, simspeciesA, simorthologs)\n",
    "\n",
    "simevents = ' '.join(simevents)\n",
    "simevents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FUS</th>\n",
       "      <th>AncChr6 AncChr20</th>\n",
       "      <th>Chr6x20 FIS</th>\n",
       "      <th>AncChr4</th>\n",
       "      <th>Chr4_1 Chr4_2 FIS</th>\n",
       "      <th>AncChr2</th>\n",
       "      <th>Chr2_1 Chr2_2 FUS</th>\n",
       "      <th>AncChr1 AncChr18</th>\n",
       "      <th>Chr1x18 TR</th>\n",
       "      <th>AncChr12 AncChr15</th>\n",
       "      <th>...</th>\n",
       "      <th>AncChr7 AncChr10</th>\n",
       "      <th>Chr7x10 FUS</th>\n",
       "      <th>AncChr14 AncChr3</th>\n",
       "      <th>Chr14x3 FIS</th>\n",
       "      <th>AncChr9</th>\n",
       "      <th>Chr9_1 Chr9_2 FUS</th>\n",
       "      <th>AncChr19 AncChr17</th>\n",
       "      <th>Chr19+17 FUS</th>\n",
       "      <th>AncChr11 AncChr8</th>\n",
       "      <th>Chr11x8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [FUS,  AncChr6 AncChr20,  Chr6x20 FIS,  AncChr4,  Chr4_1 Chr4_2 FIS,  AncChr2,  Chr2_1 Chr2_2 FUS,  AncChr1 AncChr18,  Chr1x18 TR,  AncChr12 AncChr15,  Chr12;15 Chr15;12 FUS,  AncChr7 AncChr10,  Chr7x10 FUS,  AncChr14 AncChr3,  Chr14x3 FIS,  AncChr9,  Chr9_1 Chr9_2 FUS,  AncChr19 AncChr17,  Chr19+17 FUS,  AncChr11 AncChr8,  Chr11x8]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 21 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(io.StringIO(simevents), sep = \",\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TR, AncChr2 AncChr6, Chr6;2\n"
     ]
    }
   ],
   "source": [
    "c = simevents[1]\n",
    "print(c)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.15 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "949777d72b0d2535278d3dc13498b2535136f6dfe0678499012e853ee9abcab1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
