{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import pingouin as pg\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import Orthoscripts\n",
    "\n",
    "# disable chained assignments\n",
    "pd.options.mode.chained_assignment = None "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Asterias rubens\n",
    "Astrub = Orthoscripts.readBED(\"Data/Genelists/Asterias.rubens.genelist.bed\")\n",
    "\n",
    "# Holothuria leucospilota\n",
    "Holleu = Orthoscripts.readBED(\"Data/Genelists/Holothuria.leucospilota.genelist.bed\")\n",
    "\n",
    "# Paracentrotus livides\n",
    "Parliv = Orthoscripts.readBED(\"Data/Genelists/Paracentrotus.lividus.genelist.bed\")\n",
    "\n",
    "# Branchiostoma lanceolatum\n",
    "Bralan = Orthoscripts.readBED(\"Data/Genelists/Branchiostoma.lanceolatum.genelist.bed\")\n",
    "\n",
    "# Branchiostoma floridae\n",
    "Braflo = np.loadtxt(\"Braflo_info_ra.txt\", dtype = \"str\")\n",
    "Braflo = pd.DataFrame(Braflo, columns = ['Name', 'Chromosome', 'Start', 'End', 'Dot'])\n",
    "\n",
    "# Marthasterias glacialis\n",
    "Margla = Orthoscripts.readBED(\"Data/Genelists/Marthasterias.glacialis.genelist.bed\")\n",
    "\n",
    "# Pecten maximus\n",
    "Pecmax = Orthoscripts.readBED(\"Data/Genelists/Pecten.maximus.genelist.bed\")\n",
    "\n",
    "# Stichopus chloronotus\n",
    "Stichl = Orthoscripts.readBED(\"Data/Genelists/Stichopus.chloronotus.genelist.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import orthologs\n",
    "Astrub_Holleu = np.loadtxt(\"Data/Orthologs/Asterias.rubens+Holothuria.leucospilota.txt\", dtype = \"str\")\n",
    "\n",
    "Astrub_Parliv = np.loadtxt(\"Data/Orthologs/Asterias.rubens+Paracentrotus.lividus.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Parliv = np.loadtxt(\"Data/Orthologs/Holothuria.leucospilota+Paracentrotus.lividus.txt\", dtype = \"str\")\n",
    "\n",
    "Margla_Bralan = np.loadtxt(\"Data/Orthologs/Marthasterias.glacialis+Branchiostoma.lanceolatum.txt\", dtype = \"str\")\n",
    "\n",
    "Margla_Pecmax = np.loadtxt(\"Data/Orthologs/Marthasterias.glacialis+Pecten.maximus.txt\", dtype = \"str\")\n",
    "\n",
    "Margla_Stichl = np.loadtxt(\"Data/Orthologs/Marthasterias.glacialis+Stichopus.chloronotus.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Bralan = np.loadtxt(\"Data/Orthologs/Pecten.maximus+Branchiostoma.lanceolatum.txt\", dtype = \"str\")\n",
    "\n",
    "Stichl_Bralan = np.loadtxt(\"Data/Orthologs/Stichopus.chloronotus+Branchiostoma.lanceolatum.txt\", dtype = \"str\")\n",
    "\n",
    "Stichl_Pecmax = np.loadtxt(\"Data/Orthologs/Stichopus.chloronotus+Pecten.maximus.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Holleu = np.loadtxt(\"orthology_pipeline/orthologs/Pecmax+Holleu_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Bralan = np.loadtxt(\"orthology_pipeline/orthologs/Holleu+Bralan_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Bralan = np.loadtxt(\"orthology_pipeline/orthologs/Pecmax+Bralan_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Braflo = np.loadtxt(\"orthology_pipeline/orthologs/Pecmax+Braflo_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Braflo = np.loadtxt(\"orthology_pipeline/orthologs/Holleu+Braflo_sensitive.txt\", dtype = \"str\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sorting out the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Asterias data: removing non-chromosomal values\n",
    "Astrub = Astrub.loc[Astrub['Chromosome'].str.contains('chr')]\n",
    "\n",
    "# Paracentrotus data: select chromosomal scaffolds\n",
    "Parliv_ortho = Parliv.groupby('Chromosome').size()\n",
    "Parliv_ortho = Parliv_ortho.reset_index()\n",
    "\n",
    "Parliv_ortho.columns = ['Chromosome', 'Count']\n",
    "Parliv_ortho = Parliv_ortho.loc[Parliv_ortho['Count'] >= 100]\n",
    "\n",
    "scaffolds = Parliv_ortho.Chromosome.tolist() # Remove all values from non-chromosome scaffolds\n",
    "Parliv = Parliv.loc[Parliv['Chromosome'].isin(scaffolds)]\n",
    "\n",
    "# Branchiostoma data: removing non-chromosomal values\n",
    "Bralan = Bralan.loc[Bralan['Chromosome'].str.contains('BFL_')]\n",
    "Braflo = Braflo.loc[Braflo['Chromosome'].str.contains('BFL_')]\n",
    "\n",
    "# Pecten data: Calculate number of genes on each scaffold\n",
    "Pecmax_ortho = Pecmax.groupby('Chromosome').size()\n",
    "Pecmax_ortho = Pecmax_ortho.reset_index()\n",
    "\n",
    "Pecmax_ortho.columns = ['Chromosome', 'Count']\n",
    "Pecmax_ortho = Pecmax_ortho.loc[Pecmax_ortho['Count'] >= 300]\n",
    "\n",
    "scaffolds = Pecmax_ortho.Chromosome.tolist() # Remove all values from non-chromosome scaffolds\n",
    "Pecmax = Pecmax.loc[Pecmax['Chromosome'].isin(scaffolds)]\n",
    "\n",
    "# Replace scaffold names with chromosome names\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_2', 'PYE_1')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_18', 'PYE_2')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_19', 'PYE_3')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_16', 'PYE_4')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_4', 'PYE_5')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_7', 'PYE_6')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_11', 'PYE_7')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_15', 'PYE_8')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_17', 'PYE_9')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_10', 'PYE_10')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_1', 'PYE_11')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_3', 'PYE_12')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_12', 'PYE_13')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_5', 'PYE_14')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_9', 'PYE_15')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_14', 'PYE_16')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_13', 'PYE_17')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_8', 'PYE_18')\n",
    "Pecmax = Pecmax.replace('HiC_scaffold_6', 'PYE_19')\n",
    "\n",
    "# MB orthologies: remove prefixes on gene names\n",
    "Margla_Bralan = pd.DataFrame(Margla_Bralan, columns = ['Code', 'A', 'P'])\n",
    "Margla_Bralan['P'] = Margla_Bralan['P'].str.removeprefix('Parliv_')\n",
    "Margla_Bralan = Margla_Bralan.to_numpy()\n",
    "\n",
    "# MarBra orthologies: remove suffix\n",
    "Margla_Bralan = pd.DataFrame(Margla_Bralan, columns = ['Code', 'M', 'B'])\n",
    "Margla_Bralan['M'] = Margla_Bralan['M'].str.removesuffix('.1')\n",
    "Margla_Bralan = Margla_Bralan.to_numpy()\n",
    "\n",
    "# MS orthologies: remove suffix\n",
    "Margla_Stichl = pd.DataFrame(Margla_Stichl, columns = ['Code', 'M', 'S'])\n",
    "Margla_Stichl['M'] = Margla_Stichl['M'].str.removesuffix('.1')\n",
    "Margla_Stichl['S'] = Margla_Stichl['S'].str.removesuffix('.1')\n",
    "Margla_Stichl = Margla_Stichl.to_numpy()\n",
    "\n",
    "# MP orthologies: remove suffix\n",
    "Margla_Pecmax = pd.DataFrame(Margla_Pecmax, columns = ['Code', 'M', 'P'])\n",
    "Margla_Pecmax['M'] = Margla_Pecmax['M'].str.removesuffix('.1')\n",
    "Margla_Pecmax = Margla_Pecmax.to_numpy()\n",
    "\n",
    "# AncBra orthology: \n",
    "Holleu_Bralan = pd.DataFrame(Holleu_Bralan, columns = ['Code', 'A', 'B'])\n",
    "Holleu_Bralan['B'] = Holleu_Bralan['B'].str.rsplit('_').str.get(0)\n",
    "Holleu_Bralan = Holleu_Bralan.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "def orthofy(genelistA, genelistB, orthologies):\n",
    "    \n",
    "    \"\"\"\n",
    "    inputs:\n",
    "    genelistA: gene list for species A\n",
    "    genelistB: gene list for species B\n",
    "    orthologies: orthology dataset\n",
    "    \n",
    "    outputs: dataframe with significant ortholog combinations \n",
    "             and their location in species A and B and p-Values\n",
    "    \"\"\"\n",
    "    \n",
    "    # Make ortholog dictionaries (ortholog : gene name)\n",
    "    orthdictA = dict(zip(orthologies[:, 1], orthologies[:, 0]))\n",
    "    orthdictB = dict(zip(orthologies[:, 2], orthologies[:, 0]))\n",
    "\n",
    "    # Replace genelist values with ortholog dictionary keys\n",
    "    genelistA['Name'] = genelistA['Name'].map(lambda x: orthdictA.get(x, x))\n",
    "    genelistB['Name'] = genelistB['Name'].map(lambda x: orthdictB.get(x, x))\n",
    "    \n",
    "    # Make orthology location dictionaries (ortholog : chromosome)\n",
    "    dictA = dict(zip(genelistA.loc[genelistA['Name'].str.contains('ortholog')].Name, \n",
    "                     genelistA.loc[genelistA['Name'].str.contains('ortholog')].Chromosome))\n",
    "    dictB = dict(zip(genelistB.loc[genelistB['Name'].str.contains('ortholog')].Name, \n",
    "                     genelistB.loc[genelistB['Name'].str.contains('ortholog')].Chromosome))\n",
    "    \n",
    "    # Seperate all orthology entries into new dataframe\n",
    "    AB_data = pd.DataFrame({'Orthologs': orthologies[:, 0],\n",
    "                            'A' : orthologies[:, 0],\n",
    "                            'B' : orthologies[:, 0]})\n",
    "    \n",
    "    # Replace location in A and B with ortholog location dictionary keys\n",
    "    AB_data['A'] = AB_data['A'].map(dictA)\n",
    "    AB_data['B'] = AB_data['B'].map(dictB)\n",
    "    \n",
    "    return AB_data\n",
    "\n",
    "PB = orthofy(Pecmax, Braflo, Pecmax_Braflo)\n",
    "PB = PB.dropna()\n",
    "\n",
    "# Make matrix with corresponding chromosomes\n",
    "Amp = ['BFL_11', 'BFL_10', 'BFL_16', 'BFL_8', 'BFL_3', 'BFL_1', 'BFL_18', 'BFL_14', 'BFL_15', 'BFL_5', 'BFL_7', 'BFL_3', 'BFL_17', 'BFL_3', 'BFL_19', 'BFL_12', 'BFL_1', 'BFL_13', 'BFL_2', 'BFL_2', 'BFL_6', 'BFL_9', 'BFL_4', 'BFL_4']\n",
    "Sca = ['PYE_10', 'PYE_13', 'PYE_1', 'PYE_1', 'PYE_17', 'PYE_5', 'PYE_19', 'PYE_15', 'PYE_4', 'PYE_6', 'PYE_7', 'PYE_2', 'PYE_18', 'PYE_2', 'PYE_3', 'PYE_14', 'PYE_16', 'PYE_2', 'PYE_4', 'PYE_9', 'PYE_8', 'PYE_3', 'PYE_11', 'PYE_12']\n",
    "Anc = ['G', 'B1', 'B2', 'M', 'C2', 'A1aA1b', 'B3', 'P', 'L', 'EaEb', 'F', 'QbQa', 'J1', 'QcQd', 'O2', 'N', 'A2', 'H', 'J2', 'C1', 'D', 'K', 'I', 'O1']\n",
    "ChrCorr = np.column_stack((Sca, Amp, Anc))\n",
    "\n",
    "# Make dataframe with corresponding chromosomes\n",
    "PBgenes = pd.DataFrame()\n",
    "for i in range (0, 24): \n",
    "    PBorthologs = PB.loc[(PB['A'] == ChrCorr[i, 0]) & (PB['B'] == ChrCorr[i, 1])]\n",
    "    PBorthologs['Chr'] = ChrCorr[i, 2]\n",
    "\n",
    "    PBgenes = pd.concat([PBgenes, PBorthologs])\n",
    "\n",
    "# Manually add PYE_12\n",
    "PBorthologs = PB.loc[(PB['A'] == 'PYE_12') & (PB['B'] != 'BFL_4')]\n",
    "PBorthologs['Chr'] = 'R'\n",
    "PBgenes = pd.concat([PBgenes, PBorthologs])\n",
    "\n",
    "PBgenes['BGenes'] = PBgenes.loc[:, 'Orthologs']\n",
    "PBgenes = PBgenes.rename(columns = {'Orthologs' : 'PGenes'})\n",
    "PBgenes = PBgenes[['Chr', 'A', 'PGenes', 'B', 'BGenes']]\n",
    "\n",
    "# Make reverse ortholog dictionaries (ortholog : gene name)\n",
    "orthdictA = dict(zip(Pecmax_Braflo[:, 0], Pecmax_Braflo[:, 1]))\n",
    "orthdictB = dict(zip(Pecmax_Braflo[:, 0], Pecmax_Braflo[:, 2]))\n",
    "\n",
    "# Replace values\n",
    "PBgenes['PGenes'] = PBgenes['PGenes'].map(lambda x: orthdictA.get(x, x))\n",
    "PBgenes['BGenes'] = PBgenes['BGenes'].map(lambda x: orthdictB.get(x, x))\n",
    "\n",
    "# Make dictionaries (H gene name : P/B gene name)\n",
    "orthdictP = dict(zip(Pecmax_Holleu[:, 1], Pecmax_Holleu[:, 2]))\n",
    "orthdictB = dict(zip(Holleu_Braflo[:, 2], Holleu_Braflo[:, 1]))\n",
    "\n",
    "# Replace values\n",
    "PBgenes['PGenes'] = PBgenes['PGenes'].map(lambda x: orthdictP.get(x, x))\n",
    "PBgenes['BGenes'] = PBgenes['BGenes'].map(lambda x: orthdictB.get(x, x))\n",
    "\n",
    "# Select all values orthologous in both columns\n",
    "Ancestor = PBgenes.loc[(PBgenes['PGenes'].str.contains('gene-HOLleu_')) & \n",
    "                       (PBgenes['BGenes'].str.contains('gene-HOLleu_'))]\n",
    "\n",
    "Ancestor = Ancestor.rename(columns = {'Chr' : 'Chromosome',\n",
    "                                      'PGenes' : 'Name', \n",
    "                                      'A' : 'Pchr',\n",
    "                                      'B' : 'Bchr'})\n",
    "Ancestor = Ancestor[['Chromosome', 'Name', 'Pchr', 'Bchr']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Name</th>\n",
       "      <th>Pchr</th>\n",
       "      <th>Bchr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>G</td>\n",
       "      <td>gene-HOLleu_09740</td>\n",
       "      <td>PYE_10</td>\n",
       "      <td>BFL_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>G</td>\n",
       "      <td>gene-HOLleu_10130</td>\n",
       "      <td>PYE_10</td>\n",
       "      <td>BFL_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>G</td>\n",
       "      <td>gene-HOLleu_10648</td>\n",
       "      <td>PYE_10</td>\n",
       "      <td>BFL_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>G</td>\n",
       "      <td>gene-HOLleu_11649</td>\n",
       "      <td>PYE_10</td>\n",
       "      <td>BFL_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>G</td>\n",
       "      <td>gene-HOLleu_10002</td>\n",
       "      <td>PYE_10</td>\n",
       "      <td>BFL_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6127</th>\n",
       "      <td>R</td>\n",
       "      <td>gene-HOLleu_02690</td>\n",
       "      <td>PYE_12</td>\n",
       "      <td>BFL_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6130</th>\n",
       "      <td>R</td>\n",
       "      <td>gene-HOLleu_04213</td>\n",
       "      <td>PYE_12</td>\n",
       "      <td>BFL_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6131</th>\n",
       "      <td>R</td>\n",
       "      <td>gene-HOLleu_00369</td>\n",
       "      <td>PYE_12</td>\n",
       "      <td>BFL_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6132</th>\n",
       "      <td>R</td>\n",
       "      <td>gene-HOLleu_00278</td>\n",
       "      <td>PYE_12</td>\n",
       "      <td>BFL_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6133</th>\n",
       "      <td>R</td>\n",
       "      <td>gene-HOLleu_03566</td>\n",
       "      <td>PYE_12</td>\n",
       "      <td>BFL_16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3174 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome               Name    Pchr    Bchr\n",
       "31            G  gene-HOLleu_09740  PYE_10  BFL_11\n",
       "36            G  gene-HOLleu_10130  PYE_10  BFL_11\n",
       "42            G  gene-HOLleu_10648  PYE_10  BFL_11\n",
       "53            G  gene-HOLleu_11649  PYE_10  BFL_11\n",
       "57            G  gene-HOLleu_10002  PYE_10  BFL_11\n",
       "...         ...                ...     ...     ...\n",
       "6127          R  gene-HOLleu_02690  PYE_12   BFL_5\n",
       "6130          R  gene-HOLleu_04213  PYE_12   BFL_3\n",
       "6131          R  gene-HOLleu_00369  PYE_12  BFL_11\n",
       "6132          R  gene-HOLleu_00278  PYE_12   BFL_3\n",
       "6133          R  gene-HOLleu_03566  PYE_12  BFL_16\n",
       "\n",
       "[3174 rows x 4 columns]"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ancestor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>Orthologs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_10</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_6</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>366 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A       B  Orthologs\n",
       "0    A1aA1b   BFL_1          6\n",
       "1    A1aA1b  BFL_10         74\n",
       "2    A1aA1b  BFL_11          1\n",
       "3    A1aA1b  BFL_12          3\n",
       "4    A1aA1b  BFL_13          1\n",
       "..      ...     ...        ...\n",
       "361       R   BFL_3          3\n",
       "362       R   BFL_4          7\n",
       "363       R   BFL_5          3\n",
       "364       R   BFL_6          4\n",
       "365       R   BFL_9          3\n",
       "\n",
       "[366 rows x 3 columns]"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def orthofy(genelistA, genelistB, orthologies):\n",
    "    \n",
    "    \"\"\"\n",
    "    inputs:\n",
    "    genelistA: gene list for species A\n",
    "    genelistB: gene list for species B\n",
    "    orthologies: orthology dataset\n",
    "    \n",
    "    outputs: dataframe with significant ortholog combinations \n",
    "             and their location in species A and B and p-Values\n",
    "    \"\"\"\n",
    "    \n",
    "    # Make ortholog dictionaries (ortholog : gene name)\n",
    "    orthdictA = dict(zip(orthologies[:, 1], orthologies[:, 0]))\n",
    "    orthdictB = dict(zip(orthologies[:, 2], orthologies[:, 0]))\n",
    "\n",
    "    # Replace genelist values with ortholog dictionary keys\n",
    "    genelistA['Name'] = genelistA['Name'].map(lambda x: orthdictA.get(x, x))\n",
    "    genelistB['Name'] = genelistB['Name'].map(lambda x: orthdictB.get(x, x))\n",
    "    \n",
    "    # Make orthology location dictionaries (ortholog : chromosome)\n",
    "    dictA = dict(zip(genelistA.loc[genelistA['Name'].str.contains('ortholog')].Name, \n",
    "                     genelistA.loc[genelistA['Name'].str.contains('ortholog')].Chromosome))\n",
    "    dictB = dict(zip(genelistB.loc[genelistB['Name'].str.contains('ortholog')].Name, \n",
    "                     genelistB.loc[genelistB['Name'].str.contains('ortholog')].Chromosome))\n",
    "    \n",
    "    # Seperate all orthology entries into new dataframe\n",
    "    AB_data = pd.DataFrame({'Orthologs': orthologies[:, 0],\n",
    "                            'A' : orthologies[:, 0],\n",
    "                            'B' : orthologies[:, 0]})\n",
    "    \n",
    "    # Replace location in A and B with ortholog location dictionary keys\n",
    "    AB_data['A'] = AB_data['A'].map(dictA)\n",
    "    AB_data['B'] = AB_data['B'].map(dictB)\n",
    "\n",
    "    # Calculate number of orthologs for each pair of chromosomes\n",
    "    AB_data = AB_data.groupby(['A', 'B']).count().reset_index()\n",
    "    \n",
    "    return AB_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Figure out why the function messes up chromosome correspondance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelistA = Ancestor\n",
    "genelistB = Braflo\n",
    "orthologies = Holleu_Braflo\n",
    "\n",
    "# Make ortholog dictionaries (ortholog : gene name)\n",
    "orthdictA = dict(zip(orthologies[:, 1], orthologies[:, 0]))\n",
    "orthdictB = dict(zip(orthologies[:, 2], orthologies[:, 0]))\n",
    "\n",
    "# Replace genelist values with ortholog dictionary keys\n",
    "genelistA['Name'] = genelistA['Name'].map(lambda x: orthdictA.get(x, x))\n",
    "genelistB['Name'] = genelistB['Name'].map(lambda x: orthdictB.get(x, x))\n",
    "\n",
    "# Make orthology location dictionaries (ortholog : chromosome)\n",
    "dictA = dict(zip(genelistA.loc[genelistA['Name'].str.contains('ortholog')].Name, \n",
    "                    genelistA.loc[genelistA['Name'].str.contains('ortholog')].Chromosome))\n",
    "dictB = dict(zip(genelistB.loc[genelistB['Name'].str.contains('ortholog')].Name, \n",
    "                    genelistB.loc[genelistB['Name'].str.contains('ortholog')].Chromosome))\n",
    "\n",
    "# Seperate all orthology entries into new dataframe\n",
    "AB_data = pd.DataFrame({'Orthologs': orthologies[:, 0],\n",
    "                        'A' : orthologies[:, 0],\n",
    "                        'B' : orthologies[:, 0]})\n",
    "\n",
    "# Replace location in A and B with ortholog location dictionary keys\n",
    "AB_data['A'] = AB_data['A'].map(dictA)\n",
    "AB_data['B'] = AB_data['B'].map(dictB)\n",
    "\n",
    "# Calculate number of orthologs for each pair of chromosomes\n",
    "AB_data = AB_data.groupby(['A', 'B']).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>Orthologs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_10</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1aA1b</td>\n",
       "      <td>BFL_13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_6</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>R</td>\n",
       "      <td>BFL_9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>366 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A       B  Orthologs\n",
       "0    A1aA1b   BFL_1          6\n",
       "1    A1aA1b  BFL_10         74\n",
       "2    A1aA1b  BFL_11          1\n",
       "3    A1aA1b  BFL_12          3\n",
       "4    A1aA1b  BFL_13          1\n",
       "..      ...     ...        ...\n",
       "361       R   BFL_3          3\n",
       "362       R   BFL_4          7\n",
       "363       R   BFL_5          3\n",
       "364       R   BFL_6          4\n",
       "365       R   BFL_9          3\n",
       "\n",
       "[366 rows x 3 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AB_data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.15 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "949777d72b0d2535278d3dc13498b2535136f6dfe0678499012e853ee9abcab1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
