{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import scipy.stats as stats\n",
    "import pingouin as pg\n",
    "\n",
    "import random\n",
    "from random import randrange\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import argparse\n",
    "\n",
    "import Orthoscripts\n",
    "\n",
    "pd.options.mode.chained_assignment = None "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import genelists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Asterias rubens\n",
    "Astrub = Orthoscripts.readBED(\"Data/Genelists/Asterias.rubens.genelist.bed\")\n",
    "\n",
    "# Holothuria leucospilota\n",
    "Holleu = Orthoscripts.readBED(\"Data/Genelists/Holothuria.leucospilota.genelist.bed\")\n",
    "\n",
    "# Paracentrotus livides\n",
    "Parliv = Orthoscripts.readBED(\"Data/Genelists/Paracentrotus.lividus.genelist.bed\")\n",
    "\n",
    "# Branchiostoma lanceolatum\n",
    "Bralan = Orthoscripts.readBED(\"Data/Genelists/Branchiostoma.lanceolatum.genelist.bed\")\n",
    "\n",
    "# Branchiostoma floridae\n",
    "Braflo = Orthoscripts.readBED(\"Data/Genelists/Branchiostoma.floridae.genelist.bed\", 's')\n",
    "\n",
    "# Marthasterias glacialis\n",
    "Margla = Orthoscripts.readBED(\"Data/Genelists/Marthasterias.glacialis.genelist.bed\")\n",
    "\n",
    "# Pecten maximus\n",
    "Pecmax = Orthoscripts.readBED(\"Data/Genelists/Pecmax.genelist.bed\", 's')\n",
    "\n",
    "# Stichopus chloronotus\n",
    "Stichl = Orthoscripts.readBED(\"Data/Genelists/Stichopus.chloronotus.genelist.bed\")\n",
    "\n",
    "# Amphiura filiformis \n",
    "Ampfil = Orthoscripts.readBED(\"Data/Genelists/Amphiura.filiformis.genelist.bed\")\n",
    "\n",
    "# Ephydatia muelleri\n",
    "Ephmue = Orthoscripts.readBED(\"Data/Genelists/Ephmue.genelist.bed\", 's')\n",
    "\n",
    "# Ancestor \n",
    "AniAnc = Orthoscripts.readBED(\"Data/Genelists/AniAnc.genelist.bed\", 's')\n",
    "BilAnc = Orthoscripts.readBED(\"Data/Genelists/BilAnc.genelist.bed\", 's')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import ortholog files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import orthologs\n",
    "Astrub_Holleu = np.loadtxt(\"Data/Orthologs/Asterias.rubens+Holothuria.leucospilota.txt\", dtype = \"str\")\n",
    "\n",
    "Astrub_Parliv = np.loadtxt(\"Data/Orthologs/Asterias.rubens+Paracentrotus.lividus.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Parliv = np.loadtxt(\"Data/Orthologs/Holothuria.leucospilota+Paracentrotus.lividus.txt\", dtype = \"str\")\n",
    "\n",
    "Margla_Bralan = np.loadtxt(\"Data/Orthologs/Marthasterias.glacialis+Branchiostoma.lanceolatum.txt\", dtype = \"str\")\n",
    "\n",
    "Margla_Pecmax = np.loadtxt(\"Data/Orthologs/Marthasterias.glacialis+Pecten.maximus.txt\", dtype = \"str\")\n",
    "\n",
    "Margla_Stichl = np.loadtxt(\"Data/Orthologs/Marthasterias.glacialis+Stichopus.chloronotus.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Bralan = np.loadtxt(\"Data/Orthologs/Pecten.maximus+Branchiostoma.lanceolatum.txt\", dtype = \"str\")\n",
    "\n",
    "Stichl_Bralan = np.loadtxt(\"Data/Orthologs/Stichopus.chloronotus+Branchiostoma.lanceolatum.txt\", dtype = \"str\")\n",
    "\n",
    "Stichl_Pecmax = np.loadtxt(\"Data/Orthologs/Stichopus.chloronotus+Pecten.maximus.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Holleu = np.loadtxt(\"Orthology pipeline/orthologs/Pecmax+Holleu_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Bralan = np.loadtxt(\"Orthology pipeline/orthologs/Holleu+Bralan_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Bralan = np.loadtxt(\"Orthology pipeline/orthologs/Pecmax+Bralan_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Braflo = np.loadtxt(\"Orthology pipeline/orthologs/Pecmax+Braflo_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Braflo = np.loadtxt(\"Orthology pipeline/orthologs/Holleu+Braflo_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Ampfil = np.loadtxt(\"Data/Orthologs/Holothuria.leucospilota+Amphiura.filiformis.txt\", dtype = \"str\")\n",
    "\n",
    "Braflo_Ephmue = np.loadtxt(\"Orthology pipeline/orthologs/Braflo+Ephmue_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Holleu_Ephmue = np.loadtxt(\"Orthology pipeline/orthologs/Holleu+Ephmue_sensitive.txt\", dtype = \"str\")\n",
    "\n",
    "Pecmax_Ephmue = np.loadtxt(\"Orthology pipeline/orthologs/Pecmax+Ephmue_sensitive.txt\", dtype = \"str\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sorting out the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "Astrub = Astrub.loc[Astrub['Chromosome'].str.contains('chr')]\n",
    "Bralan = Bralan.loc[Bralan['Chromosome'].str.contains('BFL_')]\n",
    "Braflo = Braflo.loc[Braflo['Chromosome'].str.contains('BFL_')]\n",
    "Pecmax = Pecmax.loc[Pecmax['Chromosome'].str.contains('PYE_')]\n",
    "Ephmue = Ephmue.loc[Ephmue['Chromosome'].str.contains('EMU_')]\n",
    "\n",
    "# Ephmue genelist: remove suffix\n",
    "Ephmue['Name'] = Ephmue['Name'].str.rsplit('.t1').str.get(0)\n",
    "\n",
    "# Parliv genelist: select chromosomal scaffolds\n",
    "Parliv = Orthoscripts.unscaff(Parliv, 100)\n",
    "Ampfil = Orthoscripts.unscaff(Ampfil, 100)\n",
    "Ephmue = Orthoscripts.unscaff(Ephmue, 600)\n",
    "\n",
    "Astrub_Parliv = Orthoscripts.orthFix(Astrub_Parliv, 'B', 'Parliv_', 1)\n",
    "Margla_Bralan = Orthoscripts.orthFix(Margla_Bralan, 'A', '.1', 0)\n",
    "Margla_Stichl = Orthoscripts.orthFix(Margla_Stichl, 'A', '.1', 0)\n",
    "Margla_Stichl = Orthoscripts.orthFix(Margla_Stichl, 'B', '.1', 0)\n",
    "Margla_Pecmax = Orthoscripts.orthFix(Margla_Pecmax, 'B', '.1', 0)\n",
    "Holleu_Ampfil = Orthoscripts.orthFix(Holleu_Ampfil, 'B', '.1', 0)\n",
    "Holleu_Bralan = Orthoscripts.orthFix(Holleu_Bralan, 'B', '_', 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### \n",
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = Orthoscripts.orthofind(Braflo, Ephmue, Braflo_Ephmue)\n",
    "Orthoscripts.orthoplot(data, 'Amphioxus', 'Sponge', 'A', 'B')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------\n",
    "\n",
    "### Test simulations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rearrangements(data):\n",
    "    # Converts table into dotplot\n",
    "    fissions = data.pivot(index = 'A', columns='B', values = 'Orthologs')\n",
    "    \n",
    "    # Picks out all rows and columns with more than one dot\n",
    "    fissions = fissions.loc[(fissions.where(fissions.isnull(), 1).sum(axis=1) > 1) | (fissions.sum(axis=0) > 1)]\n",
    "    fissions = fissions.stack(dropna = True).reset_index().groupby('A')['B'].apply(list).reset_index(name = 'B')\n",
    "    fissions['B'] = [', '.join(map(str, l)) for l in fissions['B']] # Convert list to str\n",
    "\n",
    "    # Identify all translocations\n",
    "    translocations = fissions.groupby('B').filter(lambda g: len(g) > 1)\n",
    "    # Remove all translocations from list of fissions\n",
    "    fissions = fissions[~ fissions.isin(translocations)]\n",
    "    fissions.dropna(inplace = True)\n",
    "\n",
    "    # Converts table into dotplot\n",
    "    fusions = data.pivot(index = 'B', columns = 'A', values = 'Orthologs')\n",
    "    fusions = fusions.loc[(fusions.where(fusions.isnull(), 1).sum(axis=1) > 1) | (fusions.sum(axis=0) > 1)]\n",
    "    fusions = fusions.stack(dropna = True).reset_index()\n",
    "    \n",
    "    # Picks out all rows and columns with more than one dot\n",
    "    translocations = fusions.groupby('A').filter(lambda g: len(g) > 1)\n",
    "    \n",
    "    # Remove all translocations from list of fissions\n",
    "    fusions = fusions[~ fusions.isin(translocations)]\n",
    "\n",
    "    # Identify and isolate the translocations\n",
    "    translocations = translocations.groupby('B')['A'].apply(list).reset_index()\n",
    "    translocations['A'] = [', '.join(map(str, l)) for l in translocations['A']]\n",
    "    translocations = translocations.groupby('A')['B'].apply(list).reset_index()\n",
    "    translocations['B'] = [', '.join(map(str, l)) for l in translocations['B']]\n",
    "\n",
    "    fusions = fusions.groupby('B')['A'].apply(list).reset_index(name = 'A')\n",
    "    fusions['A'] = [', '.join(map(str, l)) for l in fusions['A']]\n",
    "\n",
    "    events = []\n",
    "    for index, row in fissions.iterrows():\n",
    "        events.append(''.join(('Fission of ancestral chromosome ', row['A'], ' into ', row['B'])))\n",
    "\n",
    "    for index, row in fusions.iterrows():\n",
    "        events.append(''.join(('Fusion of ancestral chromosome ', row['A'], ' into ', row['B'])))\n",
    "        \n",
    "    for index, row in translocations.iterrows():\n",
    "        events.append(''.join(('Translocation of ancestral chromosome ', row['A'], ' into ', row['B'])))\n",
    "    \n",
    "    return events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Fission of ancestral chromosome AncChr1 into Chr1_1, Chr1_2',\n",
       " 'Fission of ancestral chromosome AncChr10 into Chr10_1, Chr10_2',\n",
       " 'Fission of ancestral chromosome AncChr18 into Chr18_1, Chr18_2',\n",
       " 'Fission of ancestral chromosome AncChr2 into Chr2_1, Chr2_2',\n",
       " 'Fusion of ancestral chromosome Chr12x9 into AncChr12, AncChr9',\n",
       " 'Fusion of ancestral chromosome Chr17x13 into AncChr13, AncChr17',\n",
       " 'Fusion of ancestral chromosome Chr3+16 into AncChr16, AncChr3',\n",
       " 'Fusion of ancestral chromosome Chr6+11 into AncChr11, AncChr6',\n",
       " 'Fusion of ancestral chromosome Chr8x7 into AncChr7, AncChr8',\n",
       " 'Translocation of ancestral chromosome AncChr14, AncChr19 into Chr14;19, Chr19;14']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rearrangements(simdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# counts the rearrangements\n",
    "for i in range(100+1):\n",
    "    input = 'Simulations/Ancestor_' + str(i) + '.bed'\n",
    "    ancestor = Orthoscripts.readBED(input, 's')\n",
    "    input = 'Simulations/SpeciesA_' + str(i) + '.bed'\n",
    "    speciesA = Orthoscripts.readBED(input, 's')\n",
    "    input = 'Simulations/Ancestor+SpeciesA_' + str(i) + '.txt'\n",
    "    orthos = np.loadtxt(input, dtype = \"str\")\n",
    "\n",
    "    data = Orthoscripts.orthologies(ancestor, speciesA, orthos)\n",
    "    \n",
    "    outfile = 'Simulations/Rearrangements_' + str(i) + '.txt'\n",
    "    Orthoscripts.rearrangements(data, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulator(Nchr = 20, Ngene = 100, Nevents = 10, Nruns = 1):\n",
    "    def makeancestor(Nchr, Ngene):\n",
    "        ancestor = pd.DataFrame(columns = ['Chromosome'])\n",
    "        for i in range(Nchr):\n",
    "            row = {'Chromosome' : (i + 1)}\n",
    "            for i in range(Ngene):\n",
    "                    ancestor = pd.concat([ancestor, pd.DataFrame([row])], ignore_index = True)\n",
    "        ancestor['Name'] = (ancestor.reset_index().index + 1)\n",
    "\n",
    "        return ancestor\n",
    "\n",
    "    # Dummy BED files :: type 'anc' for ancestor, 'des' for descendant\n",
    "    def dummyBED(genome, type):\n",
    "        if type == 'anc':\n",
    "            genome['Chromosome'] = 'AncChr' + genome['Chromosome'].astype(str)\n",
    "            genome['Name'] = 'ancg_' + genome['Name'].astype(str)\n",
    "            \n",
    "        if type == 'des':\n",
    "            genome['Chromosome'] = 'Chr' + genome['Chromosome'].astype(str)\n",
    "            genome['Name'] = 'g_' + genome['Name'].astype(str)\n",
    "        \n",
    "        genome['Start'] = np.arange(len(genome))\n",
    "        genome['End'] = np.arange(len(genome)) + 5\n",
    "        \n",
    "        genome = genome[['Chromosome', 'Start', 'End', 'Name']]\n",
    "            \n",
    "        return genome\n",
    "\n",
    "    # Dummy ortholog file\n",
    "    def dummyOrthologs(genome):\n",
    "        orthologs = pd.DataFrame()\n",
    "        \n",
    "        orthologs['Orthologs'] = np.arange(len(genome)) + 1\n",
    "        orthologs['speciesA'] = np.arange(len(genome)) + 1\n",
    "        orthologs['speciesB'] = np.arange(len(genome)) + 1\n",
    "        \n",
    "        orthologs['Orthologs'] = 'orthologs_' + orthologs['Orthologs'].astype(str)\n",
    "        orthologs['speciesA'] = 'ancg_' + orthologs['speciesA'].astype(str)\n",
    "        orthologs['speciesB'] = 'g_' + orthologs['speciesB'].astype(str)\n",
    "        \n",
    "        orthologs = orthologs.to_numpy()\n",
    "        \n",
    "        return orthologs\n",
    "\n",
    "    def mixing(genome, mixing):\n",
    "        genes = genome['Name'].to_numpy()\n",
    "        n = len(genes)\n",
    "        for i in range(int(mixing * n)):\n",
    "            g1, g2 = randrange(n), randrange(n)\n",
    "            genes[g2], genes[g1] = genes[g1], genes[g2]\n",
    "\n",
    "            genome['Name'] = genes\n",
    "            # genome['Chromosome'] = f'{fuse1}x{fuse2}'\n",
    "            \n",
    "    def fusion(genome, chr, mixing = 0):\n",
    "        '''\n",
    "        inputs: \n",
    "        ancestor : df with chromosome name | gene name\n",
    "        mixing : float between 0 and 1, where 1 implies extreme mixing and 0 implies no mixing\n",
    "        '''\n",
    "        \n",
    "        # Randomly select two chromosomes to fuse\n",
    "        A = random.choice(chr)\n",
    "        B = random.choice(chr)\n",
    "        \n",
    "        chr = [x for x in chr if x not in (A, B)]\n",
    "        \n",
    "        if A == B: # Just so the same chromosome isn't selected twice\n",
    "            B = random.choice(chr)\n",
    "\n",
    "        fusion = ancestor.loc[ancestor['Chromosome'].isin([A, B])]\n",
    "        \n",
    "        # Apply mixing if required\n",
    "        if mixing > 0:\n",
    "            genes = fusion['Name'].to_numpy()\n",
    "            n = len(genes)\n",
    "            for i in range(int(mixing * n)):\n",
    "                g1, g2 = randrange(n), randrange(n)\n",
    "                genes[g2], genes[g1] = genes[g1], genes[g2]\n",
    "\n",
    "            fusion['Name'] = genes\n",
    "            fusion['Chromosome'] = f'{A}x{B}'\n",
    "            log = f'Fusion of ancestral chromosome AncChr{A}, AncChr{B} into Chr{A}x{B}'\n",
    "            \n",
    "        else:\n",
    "            fusion['Chromosome'] = f'{A}+{B}'\n",
    "            log = f'Fusion of ancestral chromosome AncChr{A}, AncChr{B} into Chr{A}+{B}'\n",
    "        \n",
    "        # Remove the unfused chromosomes\n",
    "        genome.drop(genome[genome['Chromosome'].isin([A, B])].index, inplace = True)\n",
    "        genome = pd.concat([genome, fusion])\n",
    "        \n",
    "        return genome, log, chr\n",
    "\n",
    "    def fission(genome, chr):\n",
    "        # Randomly select a chromosome for fission\n",
    "        A = random.choice(chr)\n",
    "        fission = genome.loc[genome['Chromosome'] == A]\n",
    "        chr.remove(A)\n",
    "        \n",
    "        pos = random.choice(range(1, Ngene))\n",
    "\n",
    "        # Add the new chromosomes back into the genome\n",
    "        chr1 = fission.iloc[: pos]\n",
    "        chr1['Chromosome'] = f'{A}_1'\n",
    "        \n",
    "        chr2 = fission.iloc[pos :]\n",
    "        chr2['Chromosome'] = f'{A}_2'\n",
    "        \n",
    "        # Remove the fission chromosome from the genome\n",
    "        genome = pd.concat([genome, chr1, chr2])\n",
    "        genome = genome[genome.Chromosome != A]\n",
    "        \n",
    "        log = f'Fission of ancestral chromosome AncChr{A} into Chr{A}_1, Chr{A}_2'\n",
    "        \n",
    "        return genome, log, chr\n",
    "\n",
    "    def translocation(genome, chr):\n",
    "        # Randomly select two chromosomes for translocation\n",
    "        A = random.choice(chr)\n",
    "        B = random.choice(chr)\n",
    "        \n",
    "        if A == B: # Just so the same chromosome isn't selected twice\n",
    "            B = random.choice(chr)\n",
    "        \n",
    "        chr = [x for x in chr if x not in (A, B)]\n",
    "        \n",
    "        chrA = genome.loc[genome['Chromosome'] == A]\n",
    "        chrB = genome.loc[genome['Chromosome'] == B]\n",
    "        \n",
    "        # Randomly select two break point positions\n",
    "        posA = random.choice(range(1, Ngene))\n",
    "        posB = random.choice(range(1, Ngene))\n",
    "        \n",
    "        # Join the fragments to form recombinant chromosomes\n",
    "        chr1 = pd.concat([chrA.iloc[: posA], chrB.iloc[posB :]])\n",
    "        chr1['Chromosome'] = f'{A};{B}'\n",
    "        chr2 = pd.concat([chrB.iloc[: posB], chrA.iloc[posA :]])\n",
    "        chr2['Chromosome'] = f'{B};{A}'\n",
    "        \n",
    "        # Remove the original chromosomes from the genome\n",
    "        genome = pd.concat([genome, chr1, chr2]).drop(genome[(genome['Chromosome'] == A) & (genome['Chromosome'] == B)].index)\n",
    "        \n",
    "        log = f'Translocation of ancestral chromosomes AncChr{A}, AncChr{B} into Chr{A};{B}, Chr{B};{A}'\n",
    "        \n",
    "        return genome, log, chr\n",
    "\n",
    "    def syntenyloss(genome, chr):\n",
    "        A = random.choice(chr)\n",
    "        syn = genome.loc[genome['Chromosome'] == A]\n",
    "        genome = genome[genome.Chromosome != syn]\n",
    "        \n",
    "        chr.remove(A)\n",
    "        \n",
    "        # Assign all elements to a random chromosome\n",
    "        syn['Chromosome'] = random.choices(genome.Chromosome.unique(), k = len(syn))\n",
    "        \n",
    "        # Add back into the genome\n",
    "        genome = pd.concat([genome, syn])\n",
    "        \n",
    "        log = f'Synteny loss of AncChr{A}'\n",
    "\n",
    "        return genome, log, chr\n",
    "\n",
    "    # Apply macro-rearrangements to the ancestor\n",
    "    ancestor = makeancestor(Nchr, Ngene)\n",
    "    chr = ancestor.Chromosome.unique().tolist()\n",
    "    speciesA = ancestor.copy()\n",
    "\n",
    "    events = []\n",
    "    for event in range(Nevents):\n",
    "        r = np.random.uniform()\n",
    "        \n",
    "        if r <= 0.30:\n",
    "            if len(ancestor) < 2: continue\n",
    "            speciesA, log, chr = fission(speciesA, chr)\n",
    "            events.append(log)\n",
    "        \n",
    "        elif r <= 0.45:\n",
    "            speciesA, log, chr = translocation(speciesA, chr)\n",
    "            events.append(log)\n",
    "        \n",
    "        elif r <= 0.70:\n",
    "            speciesA, log, chr = fusion(speciesA, chr)\n",
    "            events.append(log)\n",
    "        \n",
    "        elif r <= 0.99:\n",
    "            speciesA, log, chr = fusion(speciesA, chr, mixing = 0.5)\n",
    "            events.append(log)\n",
    "            \n",
    "        else:\n",
    "            # speciesA, log, chr = syntenyloss(speciesA, chr)\n",
    "            # events.append(log)\n",
    "            # print(log)\n",
    "            continue\n",
    "        \n",
    "    ancestor = dummyBED(ancestor, 'anc')\n",
    "    speciesA = dummyBED(speciesA, 'des')\n",
    "    orthologs = dummyOrthologs(ancestor)\n",
    "    \n",
    "    return ancestor, speciesA, orthologs, events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "simancestor, simspeciesA, simorthologs, simevents = simulator()\n",
    "simdata = Orthoscripts.orthologies(simancestor, simspeciesA, simorthologs)\n",
    "algoevents = rearrangements(simdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Fusion of ancestral chromosome AncChr14, AncChr18 into Chr14+18',\n",
       " 'Fission of ancestral chromosome AncChr10 into Chr10_1, Chr10_2',\n",
       " 'Fission of ancestral chromosome AncChr12 into Chr12_1, Chr12_2',\n",
       " 'Fusion of ancestral chromosome AncChr4, AncChr20 into Chr4+20',\n",
       " 'Translocation of ancestral chromosomes AncChr11, AncChr3 into Chr11;3, Chr3;11',\n",
       " 'Fission of ancestral chromosome AncChr19 into Chr19_1, Chr19_2',\n",
       " 'Fusion of ancestral chromosome AncChr9, AncChr15 into Chr9x15',\n",
       " 'Fusion of ancestral chromosome AncChr6, AncChr13 into Chr6x13',\n",
       " 'Fusion of ancestral chromosome AncChr17, AncChr7 into Chr17x7',\n",
       " 'Fusion of ancestral chromosome AncChr1, AncChr5 into Chr1x5']"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simevents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Fission of ancestral chromosome AncChr10 into Chr10_1, Chr10_2',\n",
       " 'Fission of ancestral chromosome AncChr12 into Chr12_1, Chr12_2',\n",
       " 'Fission of ancestral chromosome AncChr19 into Chr19_1, Chr19_2',\n",
       " 'Fusion of ancestral chromosome Chr14+18 into AncChr14, AncChr18',\n",
       " 'Fusion of ancestral chromosome Chr17x7 into AncChr17, AncChr7',\n",
       " 'Fusion of ancestral chromosome Chr1x5 into AncChr1, AncChr5',\n",
       " 'Fusion of ancestral chromosome Chr4+20 into AncChr20, AncChr4',\n",
       " 'Fusion of ancestral chromosome Chr6x13 into AncChr13, AncChr6',\n",
       " 'Fusion of ancestral chromosome Chr9x15 into AncChr15, AncChr9',\n",
       " 'Translocation of ancestral chromosome AncChr11, AncChr3 into Chr11;3, Chr3;11']"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "algoevents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Translocation of ancestral chromosomes AncChr11, AncChr3 into Chr11;3, Chr3;11',\n",
       " 'Fusion of ancestral chromosome AncChr6, AncChr13 into Chr6x13',\n",
       " 'Fusion of ancestral chromosome AncChr4, AncChr20 into Chr4+20',\n",
       " 'Fusion of ancestral chromosome AncChr9, AncChr15 into Chr9x15',\n",
       " 'Fusion of ancestral chromosome AncChr1, AncChr5 into Chr1x5',\n",
       " 'Fusion of ancestral chromosome AncChr17, AncChr7 into Chr17x7',\n",
       " 'Fusion of ancestral chromosome AncChr14, AncChr18 into Chr14+18']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(simevents).difference(algoevents))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr3</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>g_201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr3</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>g_202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr3</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>g_203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr3</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>g_204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr3</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>g_205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>Chr10x2</td>\n",
       "      <td>2395</td>\n",
       "      <td>2400</td>\n",
       "      <td>g_982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>Chr10x2</td>\n",
       "      <td>2396</td>\n",
       "      <td>2401</td>\n",
       "      <td>g_943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2397</th>\n",
       "      <td>Chr10x2</td>\n",
       "      <td>2397</td>\n",
       "      <td>2402</td>\n",
       "      <td>g_998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2398</th>\n",
       "      <td>Chr10x2</td>\n",
       "      <td>2398</td>\n",
       "      <td>2403</td>\n",
       "      <td>g_945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2399</th>\n",
       "      <td>Chr10x2</td>\n",
       "      <td>2399</td>\n",
       "      <td>2404</td>\n",
       "      <td>g_908</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2400 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome  Start   End   Name\n",
       "0          Chr3      0     5  g_201\n",
       "1          Chr3      1     6  g_202\n",
       "2          Chr3      2     7  g_203\n",
       "3          Chr3      3     8  g_204\n",
       "4          Chr3      4     9  g_205\n",
       "...         ...    ...   ...    ...\n",
       "2395    Chr10x2   2395  2400  g_982\n",
       "2396    Chr10x2   2396  2401  g_943\n",
       "2397    Chr10x2   2397  2402  g_998\n",
       "2398    Chr10x2   2398  2403  g_945\n",
       "2399    Chr10x2   2399  2404  g_908\n",
       "\n",
       "[2400 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input = 'Simulations/Ancestor_' + str(4) + '.bed'\n",
    "ancestor = Orthoscripts.readBED(input, 's')\n",
    "input = 'Simulations/SpeciesA_' + str(4) + '.bed'\n",
    "speciesA = Orthoscripts.readBED(input, 's')\n",
    "input = 'Simulations/Ancestor+SpeciesA_' + str(4) + '.txt'\n",
    "orthos = np.loadtxt(input, dtype = \"str\")\n",
    "\n",
    "speciesA"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.15 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "949777d72b0d2535278d3dc13498b2535136f6dfe0678499012e853ee9abcab1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
